xl-objects.go

```diff
@@ -190,9 +190,11 @@ func (xl xlObjects) GetObject(bucket, object string, startOffset int64) (io.Read
 			// Reset offset to 0 as it would be non-0 only for the first loop if startOffset is non-0.
 			offset = 0
 			if _, err := io.Copy(fileWriter, r); err != nil {
+				r.Close()
```

Comment:
This is incorrect we need to close reader pipe with error so that the other end stops writing to the pipe and error out.

```
@@ -185,14 +186,18 @@ func (xl xlObjects) GetObject(bucket, object string, startOffset int64) (io.Read
                        r, err := xl.storage.ReadFile(bucket, pathJoin(object, partNumToPartFileName(part.PartNumber)), offset)
                        if err != nil {
                                fileWriter.CloseWithError(err)
+                               r.(*io.PipeReader).CloseWithError(err)
                                return
                        }
                        // Reset offset to 0 as it would be non-0 only for the first loop if startOffset is non-0.
                        offset = 0
                        if _, err := io.Copy(fileWriter, r); err != nil {
                                fileWriter.CloseWithError(err)
+                               r.(*io.PipeReader).CloseWithError(err)
                                return
                        }
+                       // Need to close, since we have read everything succesfully.
+                       r.Close()
                }
                fileWriter.Close()
        }()
```